File: sdk/ctv/tvos/Sources/CTVSDK/Network/NetworkMonitor.swift
import Foundation
import Network

/// NetworkMonitor for tvOS - Monitors network connectivity for CTV applications.
///
/// This class provides:
/// - Real-time network state monitoring using NWPathMonitor
/// - Fast-fail for no-network conditions (critical for CTV ad breaks)
/// - Connection quality assessment optimized for streaming
/// - Callback-based network state change notifications
@available(tvOS 12.0, *)
public final class NetworkMonitor {
    
    /// Shared instance for convenience
    public static let shared = NetworkMonitor()
    
    /// Fast-fail timeout when offline (seconds) - shorter for CTV ad breaks
    public static let offlineFastFailTimeout: TimeInterval = 0.05
    
    /// Normal network timeout (seconds)
    public static let normalTimeout: TimeInterval = 8.0
    
    /// Network state information
    public struct NetworkState: Equatable {
        public let isConnected: Bool
        public let connectionType: ConnectionType
        public let isExpensive: Bool
        public let isConstrained: Bool
        public let timestamp: Date
        
        public static let offline = NetworkState(
            isConnected: false,
            connectionType: .none,
            isExpensive: false,
            isConstrained: false,
            timestamp: Date()
        )
    }
    
    /// Connection type enumeration - tvOS typically uses ethernet or wifi
    public enum ConnectionType: String {
        case none
        case wifi
        case ethernet
        case other
    }
    
    /// Result of a pre-flight check before network operations
    public enum PreflightResult {
        /// Network is available, proceed with operation
        case proceed(NetworkState)
        
        /// Network unavailable, fast-fail with this error
        case fastFail(reason: String, state: NetworkState)
    }
    
    /// Listener closure type for network state changes
    public typealias NetworkStateListener = (NetworkState) -> Void
    
    private let monitor: NWPathMonitor
    private let queue = DispatchQueue(label: "com.rivalapex.ctv.networkmonitor", qos: .utility)
    
    private var currentState = NetworkState.offline {
        didSet {
            if oldValue != currentState {
                notifyListeners()
            }
        }
    }
    
    private var listeners: [UUID: NetworkStateListener] = [:]
    private let listenersLock = NSLock()
    
    private var isMonitoring = false
    
    /// Last connectivity change time for rate limiting reconnection attempts
    private var lastConnectivityChange: Date?
    
    public init() {
        self.monitor = NWPathMonitor()
        setupMonitor()
    }
    
    private func setupMonitor() {
        monitor.pathUpdateHandler = { [weak self] path in
            self?.updateState(from: path)
        }
    }
    
    /// Starts monitoring network state changes
    public func startMonitoring() {
        guard !isMonitoring else { return }
        isMonitoring = true
        monitor.start(queue: queue)
    }
    
    /// Stops monitoring network state changes
    public func stopMonitoring() {
        guard isMonitoring else { return }
        isMonitoring = false
        monitor.cancel()
    }
    
    /// Gets the current network state
    public var state: NetworkState {
        return currentState
    }
    
    /// Checks if network is currently connected
    public var isConnected: Bool {
        return currentState.isConnected
    }
    
    /// Performs a preflight check before network operations
    /// Returns immediately if offline, allowing fast-fail behavior
    public func preflight() -> PreflightResult {
        let state = currentState
        
        if state.isConnected {
            return .proceed(state)
        } else {
            return .fastFail(reason: "No network connection - ad break unavailable", state: state)
        }
    }
    
    /// Gets the appropriate timeout based on network state
    /// Returns fast-fail timeout when offline
    public func effectiveTimeout() -> TimeInterval {
        return isConnected ? Self.normalTimeout : Self.offlineFastFailTimeout
    }
    
    /// Adds a listener for network state changes
    /// - Returns: A token to use when removing the listener
    @discardableResult
    public func addListener(_ listener: @escaping NetworkStateListener) -> UUID {
        let token = UUID()
        listenersLock.lock()
        listeners[token] = listener
        listenersLock.unlock()
        
        // Immediately notify with current state
        listener(currentState)
        return token
    }
    
    /// Removes a listener using its token
    public func removeListener(token: UUID) {
        listenersLock.lock()
        listeners.removeValue(forKey: token)
        listenersLock.unlock()
    }
    
    /// Checks if connection is suitable for high-bitrate streaming
    /// Ethernet and stable WiFi are preferred for CTV
    public func isSuitableForStreaming() -> Bool {
        let state = currentState
        guard state.isConnected else { return false }
        
        // Ethernet is always preferred for CTV
        if state.connectionType == .ethernet {
            return true
        }
        
        // WiFi without constraints is acceptable
        return state.connectionType == .wifi && !state.isConstrained
    }
    
    /// Gets streaming quality hints for adaptive bitrate
    public func streamingQualityHints() -> [String: Any] {
        let state = currentState
        return [
            "isConnected": state.isConnected,
            "connectionType": state.connectionType.rawValue,
            "suggestedQuality": isSuitableForStreaming() ? "high" : "adaptive",
            "shouldBuffer": !isSuitableForStreaming(),
            "fastFailTimeout": Self.offlineFastFailTimeout,
            "normalTimeout": Self.normalTimeout
        ]
    }
    
    /// Gets time since last connectivity change (for rate limiting)
    public func timeSinceLastChange() -> TimeInterval? {
        guard let lastChange = lastConnectivityChange else { return nil }
        return Date().timeIntervalSince(lastChange)
    }
    
    private func updateState(from path: NWPath) {
        let connectionType: ConnectionType
        if path.usesInterfaceType(.wiredEthernet) {
            connectionType = .ethernet
        } else if path.usesInterfaceType(.wifi) {
            connectionType = .wifi
        } else if path.status == .satisfied {
            connectionType = .other
        } else {
            connectionType = .none
        }
        
        lastConnectivityChange = Date()
        
        currentState = NetworkState(
            isConnected: path.status == .satisfied,
            connectionType: connectionType,
            isExpensive: path.isExpensive,
            isConstrained: path.isConstrained,
            timestamp: Date()
        )
    }
    
    private func notifyListeners() {
        let state = currentState
        listenersLock.lock()
        let currentListeners = listeners
        listenersLock.unlock()
        
        for (_, listener) in currentListeners {
            DispatchQueue.main.async {
                listener(state)
            }
        }
    }
}

/// Error type for fast-fail on no-network conditions
public struct NoNetworkError: Error {
    public let message: String
    public let networkState: NetworkMonitor.NetworkState
    
    public init(message: String, networkState: NetworkMonitor.NetworkState) {
        self.message = message
        self.networkState = networkState
    }
}

File: sdk/ctv/android-tv/src/main/kotlin/com/rivalapexmediation/ctv/network/NetworkMonitor.kt
package com.rivalapexmediation.ctv.network

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.os.Build
import android.os.Handler
import android.os.Looper
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference

/**
 * NetworkMonitor for Android TV - Monitors network connectivity for CTV applications.
 *
 * This class provides:
 * - Real-time network state monitoring using ConnectivityManager.NetworkCallback
 * - Fast-fail for no-network conditions (critical for CTV ad breaks)
 * - Connection quality assessment optimized for streaming
 * - Callback-based network state change notifications
 */
object NetworkMonitor {
    
    /** Fast-fail timeout when offline (ms) - shorter for CTV ad breaks */
    const val OFFLINE_FAST_FAIL_TIMEOUT_MS = 50L
    
    /** Normal network timeout (ms) */
    const val NORMAL_TIMEOUT_MS = 8000L
    
    /** Network state information */
    data class NetworkState(
        val isConnected: Boolean,
        val connectionType: ConnectionType,
        val hasInternet: Boolean,
        val bandwidthKbps: Int,
        val timestamp: Long = System.currentTimeMillis()
    ) {
        companion object {
            val OFFLINE = NetworkState(
                isConnected = false,
                connectionType = ConnectionType.NONE,
                hasInternet = false,
                bandwidthKbps = 0
            )
        }
    }
    
    /** Connection type enumeration - Android TV typically uses ethernet */
    enum class ConnectionType {
        NONE,
        ETHERNET,
        WIFI,
        OTHER
    }
    
    /** Result of a pre-flight check before network operations */
    sealed class PreflightResult {
        /** Network is available, proceed with operation */
        data class Proceed(val state: NetworkState) : PreflightResult()
        
        /** Network unavailable, fast-fail with this error */
        data class FastFail(val reason: String, val state: NetworkState) : PreflightResult()
    }
    
    /** Listener interface for network state changes */
    fun interface NetworkStateListener {
        fun onNetworkStateChanged(state: NetworkState)
    }
    
    private var context: Context? = null
    private var connectivityManager: ConnectivityManager? = null
    
    private val currentState = AtomicReference(NetworkState.OFFLINE)
    private val isMonitoring = AtomicBoolean(false)
    private val listeners = ConcurrentHashMap<String, NetworkStateListener>()
    
    private var lastConnectivityChangeTime = 0L
    
    private val mainHandler = Handler(Looper.getMainLooper())
    
    private val networkCallback = object : ConnectivityManager.NetworkCallback() {
        override fun onAvailable(network: Network) {
            updateNetworkState()
        }
        
        override fun onLost(network: Network) {
            lastConnectivityChangeTime = System.currentTimeMillis()
            currentState.set(NetworkState.OFFLINE)
            notifyListeners()
        }
        
        override fun onCapabilitiesChanged(network: Network, capabilities: NetworkCapabilities) {
            updateNetworkState()
        }
    }
    
    /**
     * Initializes the NetworkMonitor with application context.
     * Should be called early in app lifecycle.
     */
    fun init(context: Context) {
        this.context = context.applicationContext
        this.connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
        
        // Get initial state
        updateNetworkState()
    }
    
    /**
     * Starts monitoring network state changes.
     */
    fun startMonitoring() {
        if (isMonitoring.getAndSet(true)) return
        
        val cm = connectivityManager ?: return
        
        val request = NetworkRequest.Builder()
            .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
            .build()
        
        try {
            cm.registerNetworkCallback(request, networkCallback)
        } catch (e: Exception) {
            isMonitoring.set(false)
        }
    }
    
    /**
     * Stops monitoring network state changes.
     */
    fun stopMonitoring() {
        if (!isMonitoring.getAndSet(false)) return
        
        try {
            connectivityManager?.unregisterNetworkCallback(networkCallback)
        } catch (e: Exception) {
            // Ignore if not registered
        }
    }
    
    /**
     * Gets the current network state.
     */
    val state: NetworkState
        get() = currentState.get()
    
    /**
     * Checks if network is currently connected.
     */
    val isConnected: Boolean
        get() = currentState.get().isConnected
    
    /**
     * Performs a preflight check before network operations.
     * Returns immediately if offline, allowing fast-fail behavior.
     */
    fun preflight(): PreflightResult {
        val state = currentState.get()
        
        return if (state.isConnected && state.hasInternet) {
            PreflightResult.Proceed(state)
        } else {
            PreflightResult.FastFail("No network connection - ad break unavailable", state)
        }
    }
    
    /**
     * Gets the appropriate timeout based on network state.
     * Returns fast-fail timeout when offline.
     */
    fun effectiveTimeoutMs(): Long {
        return if (isConnected) NORMAL_TIMEOUT_MS else OFFLINE_FAST_FAIL_TIMEOUT_MS
    }
    
    /**
     * Adds a listener for network state changes.
     * @return A token to use when removing the listener
     */
    fun addListener(listener: NetworkStateListener): String {
        val token = java.util.UUID.randomUUID().toString()
        listeners[token] = listener
        
        // Immediately notify with current state
        listener.onNetworkStateChanged(currentState.get())
        return token
    }
    
    /**
     * Removes a listener using its token.
     */
    fun removeListener(token: String) {
        listeners.remove(token)
    }
    
    /**
     * Checks if connection is suitable for high-bitrate streaming.
     * Ethernet is preferred for CTV.
     */
    fun isSuitableForStreaming(): Boolean {
        val state = currentState.get()
        if (!state.isConnected || !state.hasInternet) return false
        
        // Ethernet is always preferred for CTV
        if (state.connectionType == ConnectionType.ETHERNET) {
            return true
        }
        
        // WiFi with good bandwidth is acceptable
        // 5000 kbps is reasonable for 1080p streaming
        return state.connectionType == ConnectionType.WIFI && state.bandwidthKbps >= 5000
    }
    
    /**
     * Gets streaming quality hints for adaptive bitrate.
     */
    fun streamingQualityHints(): Map<String, Any> {
        val state = currentState.get()
        return mapOf(
            "isConnected" to state.isConnected,
            "connectionType" to state.connectionType.name,
            "bandwidthKbps" to state.bandwidthKbps,
            "suggestedQuality" to if (isSuitableForStreaming()) "high" else "adaptive",
            "shouldBuffer" to !isSuitableForStreaming(),
            "fastFailTimeoutMs" to OFFLINE_FAST_FAIL_TIMEOUT_MS,
            "normalTimeoutMs" to NORMAL_TIMEOUT_MS
        )
    }
    
    /**
     * Gets time since last connectivity change (for rate limiting).
     */
    fun timeSinceLastChangeMs(): Long {
        return if (lastConnectivityChangeTime == 0L) {
            -1L
        } else {
            System.currentTimeMillis() - lastConnectivityChangeTime
        }
    }
    
    private fun updateNetworkState() {
        val cm = connectivityManager ?: return
        
        val activeNetwork = cm.activeNetwork
        val capabilities = activeNetwork?.let { cm.getNetworkCapabilities(it) }
        
        val connectionType = when {
            capabilities == null -> ConnectionType.NONE
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) -> ConnectionType.ETHERNET
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> ConnectionType.WIFI
            else -> ConnectionType.OTHER
        }
        
        val hasInternet = capabilities?.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) == true &&
                capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
        
        val bandwidthKbps = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            capabilities?.linkDownstreamBandwidthKbps ?: 0
        } else {
            0
        }
        
        val newState = NetworkState(
            isConnected = activeNetwork != null && capabilities != null,
            connectionType = connectionType,
            hasInternet = hasInternet,
            bandwidthKbps = bandwidthKbps
        )
        
        val oldState = currentState.getAndSet(newState)
        
        if (oldState != newState) {
            lastConnectivityChangeTime = System.currentTimeMillis()
            notifyListeners()
        }
    }
    
    private fun notifyListeners() {
        val state = currentState.get()
        for ((_, listener) in listeners) {
            mainHandler.post {
                listener.onNetworkStateChanged(state)
            }
        }
    }
}

/**
 * Exception for fast-fail on no-network conditions.
 */
class NoNetworkException(
    message: String,
    val networkState: NetworkMonitor.NetworkState
) : Exception(message)
